File: sorting/quick_sort/quick_sort.py
def partition(elements: list[int], start: int, end: int) -> int:

    pivot: int = elements[end]
    swap_index: int = start - 1

    for i in range(start, end):
        if elements[i] <= pivot:
            swap_index += 1

            elements[swap_index], elements[i] = elements[i], elements[swap_index]

    swap_index += 1
    elements[swap_index], elements[end] = elements[end], elements[swap_index]

    return swap_index


def quick_sort(elements: list[int], start: int, end: int) -> list[int]:
    if start < end:
        pivot: int = partition(elements, start, end)
        quick_sort(elements, start, pivot - 1)
        quick_sort(elements, pivot + 1, end)
    return elements


if __name__ == "__main__":
    input: list[int] = [4, 9, 9, 0, 3, 4, 5, 8, 5, 3, 5, 7, 8, 6, 2]
    print(f"{input = }")
    print(f"{quick_sort(input[:], 0, len(input) - 1) = }")

File: sorting/quick_sort/QuickSort.java
import java.util.Arrays;

public class QuickSort {
    public static void main(String[] args) {
        int[] input = { 1, 3, 5, 0, 8, 4, 2 };
        print(input, "Input array:");
        print(quickSort(input.clone(), 0, input.length - 1), "quickSort:");
    }

    public static int[] quickSort(int[] elements, int start, int end) {
        if (start < end) {
            int pivot = partition(elements, start, end);
            quickSort(elements, start, pivot - 1);
            quickSort(elements, pivot + 1, end);
        }
        return elements;
    }

    private static int partition(int[] elements, int start, int end) {
        int pivot = elements[end];
        int swapIndex = start - 1;
        int temp;
        for (int i = start; i < end; i++) {
            if (elements[i] < pivot) {
                swapIndex++;

                temp = elements[swapIndex];
                elements[swapIndex] = elements[i];
                elements[i] = temp;

            }
        }

        swapIndex++;
        temp = elements[swapIndex];
        elements[swapIndex] = elements[end];
        elements[end] = temp;

        return swapIndex;
    }

    private static void print(int[] array, String text) {
        System.out.print(text + " ");
        for (int num : array) {
            System.out.print(num + " ");
        }
        System.out.println();
    }

    private static void print(int[] array) {
        for (int num : array) {
            System.out.print(num + " ");
        }
        System.out.println();
    }
}